### DISCUSSION TITLE
System performance with lots of rules

### DISCUSSION THREAD

**Discussion by noctera** [gorules/zen#161] [Q&A]
Hi, I want to create a system, where users can create rules, therefore there might be millions, if not billions custom rules. Those rules need to be checked against incoming data that comes from Kafka. Additionally there might be some static data that can be added to the rules to be checked against. So overall there will be data coming in from a stream which should be checked against all user created rules.

What is the performance of GoRules for a high number of rules. Of course I might need horizontal scaling, but is the system capable of checking single events (including static data that will be added for more in depth rules) against a big number of rules?

**Comment by ivanmiletic** [comment]
Hi @noctera would you be able to share more details on your use case? GoRules is built for real time parsing and evaluation, BRMS can handle thousands of evaluations per second on a simple infra. BRE can be optimized to handle much more with advanced caching. Do you require decision graphs or just expression language? Who would be end user? E.g. your company operators or end customers.

**Reply by noctera** [reply]
Hi @ivanmiletic 
Thanks for the reply. The end user would be normal end customers, not company users. Unfortunately I can't really tell you about the detailed use case, as I'm still in evaluating which data would be capable for this service.

But to give a general overview. I have a data pipeline where parsed information of different filetypes will go through. Users have the knowledge of the generic content of the data from pipeline and create their custom rules in order to get notifications sent. So there might be a high throughput with lots of rules created by them.

But I was reading a little bit in the Zen documentation and found out, that it is capable of evaluating in sub milliseconds time. So generally I could deploy lots of zen instances in a cluster and add that to the pipeline for parsing. Rules are stored in an elasticsearch database for example and can be stored separated by filetypes, which allows fetching only relevant rules and saves time. As this approach would be stateless, it could easily be scaled horizontally and should be capable of handling lots of rules in milliseconds. You only need the specific hardware/cluster for it. Or am I missing something?

**Reply by ivanmiletic** [reply]
Hi @noctera ,

Seems interesting. We are working on a few projects / opportunities that require low p99 with a high throughput.

If you can, please contact us on hi@gorules.io and we can continue with the discussion.

**Reply by noctera** [reply]
Hi, @ivanmiletic I will be writing you an email
